src/FakeItEasy/Configuration/IOutAndRefParametersConfiguration.cs

```diff
@@ -19,6 +20,68 @@ public interface IOutAndRefParametersConfiguration
+        /// <param name="valueProducer">A function that produces the output values.</param>
+        /// <returns>A configuration object.</returns>
+        /// <exception cref="FakeConfigurationException">
+        /// The signatures of the faked method and the <paramref name="valueProducer"/> do not match.
+        /// </exception>
+        IAfterCallSpecifiedConfiguration<TInterface> AssignsOutAndRefParametersLazily<T1, T2>(Func<T1, T2, object[]> valueProducer);
+
+        /// <summary>
+        /// Specifies a function used to produce output values for out and ref parameters.
+        /// The values should appear in the same order as the out and ref parameters in the configured call.
+        /// Any non out and ref parameters are ignored.
+        /// The function will be called each time this call is made and can return different values.
+        /// </summary>
+        /// <typeparam name="T1">Type of the first argument of the faked method call.</typeparam>
+        /// <typeparam name="T2">Type of the second argument of the faked method call.</typeparam>
+        /// <typeparam name="T3">Type of the third argument of the faked method call.</typeparam>
+        /// <param name="valueProducer">A function that produces the output values.</param>
+        /// <returns>A configuration object.</returns>
+        /// <exception cref="FakeConfigurationException">
+        /// The signatures of the faked method and the <paramref name="valueProducer"/> do not match.
+        /// </exception>
+        IAfterCallSpecifiedConfiguration<TInterface> AssignsOutAndRefParametersLazily<T1, T2, T3>(Func<T1, T2, T3, object[]> valueProducer);
+
+        /// <summary>
+        /// Specifies a function used to produce output values for out and ref parameters.
+        /// The values should appear in the same order as the out and ref parameters in the configured call.
+        /// Any non out and ref parameters are ignored.
+        /// The function will be called each time this call is made and can return different values.
+        /// </summary>
+        /// <typeparam name="T1">Type of the first argument of the faked method call.</typeparam>
+        /// <typeparam name="T2">Type of the second argument of the faked method call.</typeparam>
+        /// <typeparam name="T3">Type of the third argument of the faked method call.</typeparam>
+        /// <typeparam name="T4">Type of the fourth argument of the faked method call.</typeparam>
+        /// <param name="valueProducer">A function that produces the output values.</param>
+        /// <returns>A configuration object.</returns>
+        /// <exception cref="FakeConfigurationException">
+        /// The signatures of the faked method and the <paramref name="valueProducer"/> do not match.
+        /// </exception>
+        IAfterCallSpecifiedConfiguration<TInterface> AssignsOutAndRefParametersLazily<T1, T2, T3, T4>(Func<T1, T2, T3, T4, object[]> valueProducer);

```

Comment:
These methods were initially extension methods, but had to be moved to the interface because of the new generic type param (see the discussion [here](https://github.com/FakeItEasy/FakeItEasy/pull/804#discussion_r71857913))